docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,628 @@
+```ruby
+RSpec.describe User::UpdateUsername do
+  describe described_class::Contract, type: :model do
+    it { is_expected.to validate_presence_of(:id) }
+    it { is_expected.to validate_presence_of(:username) }
+    it do
+      is_expected.to allow_values("0userName", "USERNAME", "username", "21421341").for(:username)
+    end
+    it { is_expected.not_to allow_values("invalid-username").for(:username) }
+  end
+
+  describe ".call" do
+    subject(:result) { described_class.call(params:, **dependencies) }
+
+    fab!(:user)
+    let(:params) { { username:, id: user_id } }
+    let(:dependencies) { { guardian: } }
+    let(:guardian) { user.guardian }
+    let(:username) { "NewUsername" }
+    let(:user_id) { user.id }
+
+    context "when contract isn’t valid" do
+      let(:username) { "----" }
+
+      it { is_expected.to fail_a_contract }
+    end
+
+    context "when model is not found" do
+      let(:user_id) { 0 }
+
+      it { is_expected.to fail_to_find_a_model(:user) }
+    end
+
+    context "when current user cannot update user's username" do
+      let(:guardian) { Guardian.new }
+
+      it { is_expected.to fail_a_policy(:can_update_username) }
+    end
+
+    context "when everything’s ok" do
+      it { is_expected.to run_successfully }
+
+      it "updates user's username" do
+        expect { result }.to change { user.reload.username }.to(username)
+      end
+
+      it "logs the action" do
+        expect { result }.to change { UserHistory.count }.by(1)
+      end
+    end
+  end
+end
+```
+First, and because a contract is present, we test it using a dedicated `describe` block. Since a contract uses `ActiveModel` under the hood, the simplest way to test it is to use [Shoulda Matchers](https://github.com/thoughtbot/shoulda-matchers).
+
+Then, we use a `describe` block for the `.call` method, which is how the service is run. We’re using a `context` for each possible branching. It’s quite easy as we just have to follow the steps we defined in the service. You can see we’re not testing all the possible values to have the contract fail: that’s because it’s tested extensively above, so here we’re just ensuring the `params` step is properly called and if a bad value is provided, then it will stop the execution of the service.
+For the other steps, if they can fail, then they should have a context using a dedicated matcher.
+
+The `run_successfully` matcher ensures the service succeeded (`result.success?` is `true`) and will provide some debugging information if that’s not the case.
+
+In the event of a matcher failing, it will output details about the result object to help debugging things:
```

Comment:
😮 very nice!

I wonder, is there an easy way to access the same output manually? I imagine that would be useful to debug issues outside of RSpec tests?

Reply:
Aha, I just got to the 'inspect_steps' section below 🎉